### PR TITLE
HOTFIX: fix for standby tasks using batching restore

### PR DESCRIPTION
**Pull request by bbejeck**

### CODE REVIEW

**Comment by bbejeck** [comment]
@enothereska here's fix for the errors exposed from the `StreamsSmokeTest.test_streams`

\cc @dguy @guozhangwang

**Comment by enothereska** [comment]
Thanks @bbejeck could we put through branch builder to test (unless you have already tested). Thanks.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6577/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6562/
Test PASSed (JDK 8 and Scala 2.12).

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 1]
```diff
@@ -204,16 +206,17 @@ public void register(final StateStore store,
...
-        final StateRestoreCallback restoreCallback = restoreCallbacks.get(storePartition.topic());
+        final BatchingStateRestoreCallback restoreCallback = getBatchingRestoreCallback(restoreCallbacks.get(storePartition.topic()));
 
         long lastOffset = -1L;
         int count = 0;
         for (final ConsumerRecord<byte[], byte[]> record : records) {
             if (record.offset() < limit) {
                 try {
-                    restoreCallback.restore(record.key(), record.value());
+                    restoreRecords.add(KeyValue.pair(record.key(), record.value()));
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: Is the `try-catch` still needed here? And is it possible that some `Exception` be thrown from `restoreCallback.restoreAll` below? If yes we need to think if we want to handle it here, or in higher-levels of the call trace.

**Comment by bbejeck** [comment]
results from branch builder test run
```
17:34:05 ================================================================================
17:34:05 SESSION REPORT (ALL TESTS)
17:34:05 ducktape version: 0.6.0
17:34:05 session_id:       2017-08-04--001
17:34:05 run time:         51 minutes 33.816 seconds
17:34:05 tests run:        29
17:34:05 passed:           25
17:34:05 failed:           0
17:34:05 ignored:          4
17:34:05 ================================================================================
........
17:35:30 {"success": false}Finished: SUCCESS
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java [hunk 2]
```diff
@@ -509,4 +556,33 @@ public void shouldDeleteCheckpointFileOnCreationIfEosEnabled() throws Exception
...
+            },
+            changelogReader,
+            false);
+    }
+
+    private MockStateStoreSupplier.MockStateStore getPersistentStore() {
+        return new MockStateStoreSupplier.MockStateStore("persistentStore", true);
+    }
+
+    private static class MockBatchingRestoreCallback extends AbstractNotifyingBatchingRestoreCallback {
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java]: We have this in `CompositeRestoreListenerTest` as well, also there is a similar `MockNoListenBatchingStateRestoreCallback` in that class as well. Could we collapse them all into a single class (we already have a `MockStateRestoreListener` in `o.a.k.test`, probably can just extend to another class under `o.a.k.test`)?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 1 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: A `RocksDBException` can be thrown during the restoration process.  The current error handing flow is we catch it in the `RocksDBStateStore`, and wrap it in a `ProcessorStateException` exception and re-throw.  We'll catch it again here, add some contextual information and re-throw for higher levels of the call handle it.  If we want to re-consider how we handle these restoration errors IMHO, it shouldn't block this PR, but add a follow on one to explore our options.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java (hunk 2 above)
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java]: Collapsed the mock objects in `CompositeRestoreListenerTest` and `ProcessorStateManagerTest` into a single class in `o.a.k.test.MockBatchingStateRestoreListener`, but we still need the `MockNoListenBatchingStateRestoreCallback` as we need a callback that does not implement the `StateRestoreListener` for testing.

**Comment by bbejeck** [comment]
@guozhangwang @enothereska updated per comments

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6613/
Test PASSed (JDK 7 and Scala 2.11).

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java (hunk 2 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java]: Sounds good.

**Comment by guozhangwang** [comment]
LGTM. Merged to trunk.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6598/
Test PASSed (JDK 8 and Scala 2.12).

**Pull request closed without merging**